apps/networkmonitor/networkmonitor.nim

```diff
@@ -82,43 +99,68 @@ proc setConnectedPeersMetrics(discoveredNodes: seq[Node],
     let ip = $typedRecord.get().ip.get().join(".")
     allPeers[peerId].ip = ip
 
-    let peer = toRemotePeerInfo(discNode.record)
-    if not peer.isOk():
-      warn "error converting record to remote peer info", record=discNode.record
-      continue
+    # try to ping the peer
+    if getTime().toUnix() >= allPeers[peerId].lastTimeConnected + ReconnectTime and allPeers[peerId].retries < MaxConnectionRetries:
+      if allPeers[peerId].retries > 0:
+        warn "trying to dial failed peer again", peerId=peerId, retry=allPeers[peerId].retries
 
-    # try to connect to the peer
-    # TODO: check last connection time and if not > x, skip connecting
-    let timedOut = not await node.connectToNodes(@[peer.get()]).withTimeout(timeout)
-    if timedOut:
-      warn "could not connect to peer, timedout", timeout=timeout, peer=peer.get()
-      # TODO: Add other staates
-      allPeers[peerId].connError = "timedout"
-      continue
+      var pingDelay:chronos.Duration
+
+      proc ping(): Future[Result[void, string]] {.async, gcsafe.} =
+        try:
+          let conn = await node.switch.dial(peerInfo.peerId, peerInfo.addrs, PingCodec)
+          pingDelay = await node.libp2pPing.ping(conn)
+          return ok()
+
+        except CatchableError:
+          var msg = getCurrentExceptionMsg()
+          if msg == "Future operation cancelled!":
+            msg = "timedout"
+          warn "failed to ping the peer", peer=peerInfo, err=msg
+
+          allPeers[peerId].connError = msg
+          return err("could not ping peer: " & msg)
+
+      let timedOut = not await ping().withTimeout(timeout)
```

Comment:
dont fully understand `proc ping(): Future[Result[void, string]] {.async, gcsafe.} =`

and this `let timedOut = not await ping().withTimeout(timeout)`


isnt something like this better to handle the timeout?
```
if not (waitFornode.libp2pPing.ping(conn)).withTimeout(timeout)):
   xxx
```

and if it errors before timeout, that is handled in the `except`?

Reply:
Maybe:D This was the first thing that came to mind when trying to overcome the weirdness of withTimeout which would not return any result from ping.

I'll rewrite it

Reply:
Feels to me that this inner `ping()` function may be unnecessary and you can just dial and await the `libp2pPing.ping()` directly? May be missing something here.

Reply:
Ok, so looking at the code again, my problem was that I need the `Connection` from `dial()` to be able to call the `ping`, but without `withTimeout` on the `dial` it will hang for some nodes. But then if I add `withTimeout` I don't see a way to extract the returned `Connection`.

This is why I wrapped it in a `proc` and just set the variables directly from inside of the inner proc. If there is another (better) way to do it, I am happy to refactor, but I don't see

cc @jm-clius